### DISCUSSION TITLE
add_redact_annot() with replace text and new font showing, "Exception: need font file or buffer"

### DISCUSSION THREAD

**Discussion by anooplsankar** [pymupdf/PyMuPDF#3396] [Looking for help]
### Description of the bug

Below is the code I wrote to replace a specific text on a PDF using **add_redact_annot()**. Everything works fine if I simply call **add_redact_annot()** without the parameters for replacing text, font, then use **insert_textbox()**. I learned from the documentation that **apply_redactions()** will internally call **insert_texbox()**, so I specified the replaced text, font name, and font size on **add_redact_annot()**. However, it throws an error on the line **apply_redactions()**, saying "Exception: need font file or buffer".

Another attempt: I am using VS Code for coding. When I type, the editor suggests **_apply_redactions()** instead of **apply_redactions()**. So, I tried with **_apply_redactions()**. This time, there are no errors, but on the PDF the texts are successfully removed, yet not replaced.

```
import fitz # PyMuPDF

doc = fitz.open("/var/www/html/pdfedit_header_footer/ist_pdf_cpy.pdf")  
page = doc[0]

# Rectangle cordinates for replacing the text
rec = [74, 93, 289, 106]

# Setting font
fnt = "/var/www/html/pdfedit_header_footer/fonts/MinionPro-Bold.otf"
page.insert_font(fontfile=fnt, fontname="F0")

# Redact annotation
page.add_redact_annot(rec, text="Replace text", fontname="F0", fontsize=20, text_color=(0, 0, 0))

# Appy redaction
page.apply_redactions(images=0, graphics=0)

doc.subset_fonts()
doc.save("redact_pdf.pdf")

doc.close()
```
Here is the complete Traceback
```
Traceback (most recent call last):
  File "/var/www/html/pdfedit_header_footer/redact_issue.py", line 20, in <module>
    page.apply_redactions(images=0, graphics=0)
  File "/home/anoop/.local/lib/python3.10/site-packages/fitz/utils.py", line 4280, in apply_redactions
    rc = shape.insert_textbox(
  File "/home/anoop/.local/lib/python3.10/site-packages/fitz/utils.py", line 3846, in insert_textbox
    xref = self.page.insert_font(
  File "/home/anoop/.local/lib/python3.10/site-packages/fitz/__init__.py", line 9166, in insert_font
    val = self._insertFont(fontname, bfname, fontfile_str, fontbuffer, set_simple, idx,
  File "/home/anoop/.local/lib/python3.10/site-packages/fitz/__init__.py", line 7964, in _insertFont
    value = JM_insert_font(pdf, bfname, fontfile,fontbuffer, set_simple, idx, wmode, serif, encoding, ordering)
  File "/home/anoop/.local/lib/python3.10/site-packages/fitz/__init__.py", line 16241, in JM_insert_font
    RAISEPY(MSG_FILE_OR_BUFFER, PyExc_ValueError)
  File "/home/anoop/.local/lib/python3.10/site-packages/fitz/__init__.py", line 17856, in RAISEPY
    raise Exception( msg)
Exception: need font file or buffer
```

### How to reproduce the bug

 Using the above code.

### PyMuPDF version

1.24.1

### Operating system

Linux

### Python version

3.10

**Comment by JorjMcKie** [comment]
You did not follow the instructions given in the documentation for the case of non-Base14 fonts. Inserting the font is not sufficient, you must also insert some text with your font.

**Comment by anooplsankar** [comment]
Thanks for the reply. Apologies, but I'm new to PyMuPDF.

Regarding your suggestion to "insert some text with your font," could you please clarify what this means? I've already mentioned that inserting text works fine with insert_textbox(). Initially, I followed your answer from https://stackoverflow.com/a/76021270/10727661, and later I switched to add_redact_annot() because my requirement is to replace text, and i know it internally calls insert_textbox(). 

Please help to fix this, or point me to the documentation where it helps to add new non-Base14 fonts.

**Reply by JorjMcKie** [reply]
converting this post to a Discusion item ...

**Comment by JorjMcKie** [comment]
This is the docu comment I was referring to:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/79573993-39af-4115-980f-c1c80ea7f9fd)

I don't know what you ultimately want to achieve. As you are insisting on your own font I tend to assume that you want to look the result as nice as possible.
**_This cannot be achieved_** by letting the method `add_redact_annot` do the insertion. Several things may go wrong like imprecise insertion point, font size differences, etc.
You should use separate steps for (1) adding and applying the annotation(s), and (2) inserting the new text.

**Comment by anooplsankar** [comment]
> This cannot be achieved by letting the method add_redact_annot do the insertion. Several things may go wrong like imprecise insertion point, font size differences, etc.

Yes it is. 
For example:
This is the output I get when I print all the fonts on the page:
```
[(67, 'pfa', 'Type1', 'BKHSHV+MinionPro-It', 'F120', ''), 
(68, 'pfa', 'Type1', 'ZQOXDL+MinionPro-Regular', 'F122', ''), 
(69, 'pfa', 'Type1', 'MPMPMB+MTSYN', 'F28', ''), 
(71, 'pfa', 'Type1', 'LZTJLX+MinionPro-Bold', 'F19', ''), 
(72, 'pfa', 'Type1', 'VAGJFV+NimbusRomNo9L-Regu', 'F126', ''), 
(74, 'pfa', 'Type1', 'MINVGI+NimbusSanL-Bold', 'F127', ''), 
(75, 'pfa', 'Type1', 'YTOBJZ+NimbusSanL-ReguItal', 'F128', ''), 
(78, 'pfa', 'Type1', 'TNUAZC+NimbusSanL-BoldItal', 'F129', ''), 
(79, 'pfa', 'Type1', 'USDBTS+NimbusSanL-Regu', 'F130', ''), 
(85, 'pfa', 'Type1', 'BCWOXX+MinionPro-BoldIt', 'F119', ''), 
(86, 'pfa', 'Type1', 'SCOOQG+FuturaStd-BookOblique', 'F125', '')]
```
So, I utilized the font F19. It displays the text with some cross boxes interspersed, but the font size does not seem to be functioning correctly.
This was the original PDF section:
![image](https://github.com/pymupdf/PyMuPDF/assets/8655070/a8ead90e-aed4-4506-9b11-6abf2bf564cd)

This below is change I get after applying add_redact_annot()
 ``` page.add_redact_annot(rec, text="Effects of modification", align=2, fontname="F19", fontsize=20, fill=(0.969, 0.969, 0.969)) ```
![image](https://github.com/pymupdf/PyMuPDF/assets/8655070/42352a15-1d63-4d13-9a8c-a53f532f716d)

So I attempted to insert the font again and utilize it, but this resulted in the error that I explained in my initial post.

> You should use separate steps for (1) adding and applying the annotation(s), and (2) inserting the new text.

Thank you, I will try this way.

**Comment by JorjMcKie** [comment]
Here is a script that does a similar thing 
[replacer2.zip](https://github.com/pymupdf/PyMuPDF/files/15023559/replacer2.zip)

**Reply by anooplsankar** [reply]
Thank you.

Got KeyError, then I apply some small changes in the function get_fontlist(). Got the output, but still showing some font issue.
![image](https://github.com/pymupdf/PyMuPDF/assets/8655070/a3914b30-70a2-496e-a47e-0551a7870d03)

Here is the PDF and the script
[text_replace.zip](https://github.com/pymupdf/PyMuPDF/files/15025192/text_replace.zip)

**Reply by anooplsankar** [reply]
Even though it's not a proper solution, I've noticed that when I use insert_font() before insert_text(), everything works fine.
```
    # Setting font
    fnt = f"fonts/{font}.otf"
    xref = page.insert_font(fontfile=fnt, fontname="F0")
    ff = doc.extract_font(xref)  # exract font buffer
    font_obj = fitz.Font(fontbuffer=ff[-1])

    # IMPORTANT: matrix to stretch or shrink new text horizontally
    matrix = adjust_matrix(font_obj, s["bbox"], text, fsize)
    page.insert_text(
        point,
        text,
        fontname="F0",
        fontsize=fsize,
        color=(1, 0, 0),
        morph=(point, matrix),  # this will shrink / stretch horizontally
    )
```
![image](https://github.com/pymupdf/PyMuPDF/assets/8655070/1b272d17-1130-48a0-a346-9bb55ec53882)
